### PR TITLE
First stab at MAINTAINERS.md & OWNERS.md & CONTRIBUTING.md

### PR DESCRIPTION
**Pull request by oschaaf**
- A pre-merge checklist
- Documents what goes into an update to the Envoy dependency

Fixes https://github.com/envoyproxy/nighthawk/issues/230

Signed-off-by: Otto van der Schaaf <oschaaf@we-amp.com>

### CODE REVIEW

**Comment by caniszczyk** [comment]
Where is the actual list of maintainers for the project?

**Comment by oschaaf** [comment]
Possibly I am using the wrong terminology here; this document targets to be helpful to [reviewers](https://github.com/orgs/envoyproxy/teams/nighthawk-reviewers).

File: MAINTAINERS.md [hunk 1]
```diff
@@ -0,0 +1,18 @@
+# Maintainers
+
+This document aims to assist maintainers.
+
```
Comment (htuch) [inline MAINTAINERS.md]: Can you add some stuff on how to spot which changes are Nighthawk local vs. those that require deeper Envoy internals and architecture knowledge?

**Comment by htuch** [comment]
@caniszczyk good point. Can we have an OWNERS.md file @oschaaf in the same style as https://github.com/envoyproxy/envoy/blob/master/OWNERS.md?

We should also link to the @envoyproxy/nighthawk-maintainers team page.

**Comment by oschaaf** [comment]
@envoyproxy/nighthawk-maintainers please check the newly created `OWNERS.md` in this PR; it has email addresses and suggests areas of expertise indicating what should be routed to whom; both need feedback.

File: MAINTAINERS.md (hunk 1 above)
Comment (oschaaf) [reply MAINTAINERS.md]: done

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+
+* Coding style mirrors [Envoy's policy](https://github.com/envoyproxy/envoy/blob/master/STYLE.md)
+
+# Breaking change policy
+
+Both API and implementation stability are important to Nighthawk. Since the API is consumed by clients beyond Nighthawk, breaking changes to that are prohibited.
+
+# Submitting a PR
+
+* Fork the repo.
```
Comment (htuch) [inline CONTRIBUTING.md]: For any verbose section, if it's the same as Envoy, just link back to the relevant anchor link in the Envoy docs. If there's a small delta, then feel free to link and state the difference.

File: OWNERS.md [hunk 3]
```diff
@@ -0,0 +1,12 @@
+* See [CONTRIBUTING.md](CONTRIBUTING.md) for general contribution guidelines.
+
+This page lists all active maintainers and their areas of expertise. This can be used for
+routing PRs, questions, etc. to the right place.
+
+# Maintainers
+
+* Otto van der Schaaf ([oschaaf](https://github.com/oschaaf)) (oschaaf@we-amp.com)
+  * All things Nighthawk
+* Harvey Tuch ([htuch](https://github.com/htuch)) (htuch@google.com)
```
Comment (htuch) [inline OWNERS.md]: You can list me as "Envoy internals, architecture and performance methodology." (at least those are the things I care about, I won't claim too much expertise in any one ;). I think @jplevyak should also have "Envoy internals and architecture" listed, as he can also provide some review skills here.

File: OWNERS.md [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+* John Plevyak ([jplevyak](https://github.com/jplevyak)) (jplevyak@google.com)
+* Jakub Sobon ([jplevyak](https://github.com/jplevyak)) (???@google.com)
```
Comment (htuch) [inline OWNERS.md]: This references jpleyvak for Jakub (@mum4k).

**Comment by oschaaf** [comment]
Thanks for the feedback @htuch, based on that I pushed 009ab22

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -0,0 +1,37 @@
+We welcome contributions from the community. Please read the following guidelines carefully to
+maximize the chances of your PR being merged.
+
+# Communication
+
+* Before starting work on a major feature, please reach out to us via GitHub, Slack,
```
Comment (mum4k) [inline CONTRIBUTING.md]: Should we include link to or at least mention the Slack workspace we are using?

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+* Coding style mirrors [Envoy's policy](https://github.com/envoyproxy/envoy/blob/master/STYLE.md)
+
+# Breaking change policy
+
+Both API and implementation stability are important to Nighthawk. Since the API is consumed by clients beyond Nighthawk, breaking changes to that are prohibited.
+
+# Submitting a PR
+
+* Generally Nighthawk mirrors [Envoy's policy](https://github.com/envoyproxy/envoy/blob/master/CONTRIBUTING.md#pr-review-policy-for-maintainers) with respect to PR submission policy.
+* Any PR that changes user-facing behavior **must** have associated documentation in [docs](docs) as
```
Comment (mum4k) [inline CONTRIBUTING.md]: (nit) The mentioned docs directory is suspiciously empty, do we have any examples that we can include, or are there some coming shortly? Or is it maybe that we didn't change any behavior yet?

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+  well as [release notes](docs/root/intro/version_history.rst).
```
Comment (mum4k) [inline CONTRIBUTING.md]: This file doesn't seem to exist, while a version_history.md directly under root does:

https://github.com/envoyproxy/nighthawk/blob/master/docs/root/version_history.md

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+
+* Coding style mirrors [Envoy's policy](https://github.com/envoyproxy/envoy/blob/master/STYLE.md)
+
+# Breaking change policy
+
+Both API and implementation stability are important to Nighthawk. Since the API is consumed by clients beyond Nighthawk, breaking changes to that are prohibited.
+
+# Submitting a PR
+
+* Generally Nighthawk mirrors [Envoy's policy](https://github.com/envoyproxy/envoy/blob/master/CONTRIBUTING.md#pr-review-policy-for-maintainers) with respect to PR submission policy.
```
Comment (mum4k) [inline CONTRIBUTING.md]: This link seems to be mentioned twice. Maybe we wanted this one in here?

https://github.com/envoyproxy/envoy/blob/master/CONTRIBUTING.md#submitting-a-pr

File: MAINTAINERS.md [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+- Does the PR have breaking changes? Then that should be explicitly mentioned in the [version history](docs/root/version_history.md).
+- New features should be added to the [version history](docs/root/version_history.md).
+- Breaking changes to the [proto apis](api/) are not allowed.
+
+## Updates to the Envoy dependency
+
+We try to regularly synchronize our Envoy dependency with the latest revision. When we do so, that looks like:
+
+- A change to [repositories.bzl](bazel/repositories.bzl) to update the commit and SHA.
+- A sync of [.bazelrc](.bazelrc)
```
Comment (mum4k) [inline MAINTAINERS.md]: Can we expand this a bit indicating the what / how? I am presuming this means copy the .bazelrc from Envoy? If yes, can we include a link? If there is more to it, can we document it?

File: MAINTAINERS.md [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+- A sync of the build image sha used in the [ci configuration](.circleci/config.yml).
```
Comment (mum4k) [inline MAINTAINERS.md]: Same comment applies here.

File: MAINTAINERS.md [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+- Any changes required to Nighthawk
```
Comment (mum4k) [inline MAINTAINERS.md]: Can we expand what we mean by this?

File: OWNERS.md [hunk 12]
```diff
@@ -0,0 +1,14 @@
...
+
+# Maintainers
+
+* Otto van der Schaaf ([oschaaf](https://github.com/oschaaf)) (oschaaf@we-amp.com)
+  * All things Nighthawk
+* Harvey Tuch ([htuch](https://github.com/htuch)) (htuch@google.com)
+  * Envoy internals, architecture and performance methodology.
+* John Plevyak ([jplevyak](https://github.com/jplevyak)) (jplevyak@google.com)
+  * Envoy internals and architecture.
+* Jakub Sobon ([mum4k](https://github.com/mum4k)) (???@google.com)
```
Comment (mum4k) [inline OWNERS.md]: The email is: mumak@google.com

For now you can list me on:
  *  APIs, general functionality and miscellany.

File: CONTRIBUTING.md (hunk 6 above)
Comment (oschaaf) [reply CONTRIBUTING.md]: Filed https://github.com/envoyproxy/nighthawk/issues/241 and self-assigned

File: CONTRIBUTING.md [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+
+# Submitting a PR
+
+* Generally Nighthawk mirrors [Envoy's policy](https://github.com/envoyproxy/envoy/blob/master/CONTRIBUTING.md#submitting-a-pr) with respect to PR submission policy.
+* Any PR that changes user-facing behavior **must** have associated documentation in [docs](docs) as
+  well as [release notes](docs/root/version_history.rst).
+
+# PR review policy for maintainers
+
+* Generally Nighthawk mirros [Envoy's policy](https://github.com/envoyproxy/envoy/blob/master/CONTRIBUTING.md#pr-review-policy-for-maintainers) with respect to maintainer review policy.
```
Comment (htuch) [inline CONTRIBUTING.md]: Nit: mirrors

File: CONTRIBUTING.md [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+* See [OWNERS.md](OWNERS.md) for the current list of maintainers.
+
+# DCO: Sign your work
+
+Commits need to be signed off. [See here](https://github.com/envoyproxy/envoy/blob/master/CONTRIBUTING.md#dco-sign-your-work).
```
Comment (htuch) [inline CONTRIBUTING.md]: Nit: I would put the See outside of the link

File: MAINTAINERS.md [hunk 15]
```diff
@@ -0,0 +1,28 @@
+# Maintainers
+
+This document aims to assist [maintainers](OWNERS.md).
+
+## Envoy domain expertise
+
+As a guideline, concepts in Nighthawk that are derived from Envoy
+require someone with Envoy domain expertise in review. Notable examples
+are the way Nighthawk internal computes cluster configuration, its
```
Comment (htuch) [inline MAINTAINERS.md]: Nit: "Nighthawk internally"

File: MAINTAINERS.md [hunk 16]
```diff
@@ -0,0 +1,28 @@
...
+connection pool derivations, the `StreamDecoder` class, as well as anything related to the Nighthawk test server.
+
+See [OWNERS.md](OWNERS.md) to find maintainers with expertise of
+Envoy internals.
+
+## Pre-merge checklist
+
+- Does the PR have breaking changes? Then that should be explicitly mentioned in the [version history](docs/root/version_history.md).
+- New features should be added to the [version history](docs/root/version_history.md).
+- Breaking changes to the [proto apis](api/) are not allowed.
```
Comment (htuch) [inline MAINTAINERS.md]: Nit: "protobuf APIs"

**Comment by oschaaf** [comment]
@htuch nits fixed in 7ff2dd22423b4c9e879b445c37fc507365c4bb94

File: CONTRIBUTING.md [hunk 17]
```diff
@@ -18,18 +16,18 @@ Both API and implementation stability are important to Nighthawk. Since the API
 
 # Submitting a PR
 
-* Generally Nighthawk mirrors [Envoy's policy](https://github.com/envoyproxy/envoy/blob/master/CONTRIBUTING.md#pr-review-policy-for-maintainers) with respect to PR submission policy.
+* Generally Nighthawk mirrors [Envoy's policy](https://github.com/envoyproxy/envoy/blob/master/CONTRIBUTING.md#submitting-a-pr) with respect to PR submission policy.
 * Any PR that changes user-facing behavior **must** have associated documentation in [docs](docs) as
-  well as [release notes](docs/root/intro/version_history.rst). 
+  well as [release notes](docs/root/version_history.rst).
```
Comment (mum4k) [inline CONTRIBUTING.md]: We may still have a wrong file extension here (.rst vs .md).

Comment (htuch) [reply CONTRIBUTING.md]: GitHub developer docs are MD, end user docs are RST.

Comment (mum4k) [reply CONTRIBUTING.md]: Thanks for the clarification, the comment was based on the fact that the linked file doesn't exist.

@oschaaf can we create the missing doc as necessary?

Comment (oschaaf) [reply CONTRIBUTING.md]: I fixed the bad link in 63710cf9543cd470141e1317e2adda5e8bd8ba82, it was some bad copying.
The file currently is MD, I was not aware of this dev/user distinction. I punted renaming this file to #244 - I will pick this up once all the doc PR's have landed.

**Pull request merged**